### PR TITLE
WIP: First cut at applicationConfig

### PR DESCRIPTION
**Pull request by bk3c**
(see #453 for context)

Ok, this is my first very rough cut at an applicationConfig object, which was just getting a little too long to put in issue comments. I will not be upset if the code itself never makes it anywhere.

Some notes (in addition to the inline file comments):

- I took James's skip_step idea and refactored it up to the conductor so the steps themselves don't have any of that logic in them, which I think is the goal. Hopefully I didn't mangle the intent of the idea.

- I think we may want all form steps to end up submitting back to the same URL across the board, representing the conductor/config component logic. That would provide a central place for the form content to be both saved (as it is now from `ApplicationConductor.sync()`) but also to put the conductor in charge of evaluating any conditionals and then sending the user to the correct next step.

- Trying to get the conditional logic completely embedded in JSON seems both difficult to do elegantly and I'm not sure it buys us that much. It's currently in there as a hybrid approach, but I don't love it.

- I still think being able to pass the equivalent of props into a given screen is important, for example the age limit in my example here. It might even be a more complex object like an AMI chart. It definitely doesn't have to be props per se -- I defer to basically everyone else on what the best way to pass the data in is, but I do think we're better off pushing the right data into the screen rather than embedding the logic in the screen to know where to go read it from. I might be being too idealistic on this point, and I guess if we just make the config available to the screen component and let it read from that context, it's not the end of the world.

- I was thinking about nextjs-like ways to handle the routing and still have the conductor really be able to evaluate things on every click, and one idea I had was that we could have a url like `/application/{configId}/step/{stepName}` and then have `/application/*` always resolve to the conductor which did the right thing based on the inputs, sort of like the listing page based on the listingId. I'm wholly unconvinced that's the right approach, but it seemed at least worth talking about.

### CODE REVIEW

**Comment by jaredcwhite** [comment]
This looks like a good first step to me. I imagine we'll probably roll this out in stages so some of the more advanced routing concepts may need to be fleshed out later. One thing that's been nagging at me is the way to express conditionals. You put in `userIsLoggedIn()`, which is logical, but my concern with that approach is we probably don't want to eval code over the wire as that could be a security risk.

Maybe the conductor could offer a selection of "condition resolvers" (not sure what the technical term is), and the value from the backend is a key. So it could instead be "userIsLoggedIn" (not a function per se), and that key in the condition resolvers set would resolve to executing the actual function "userIsLoggedIn". But some other key like "doSomethingNefarious" wouldn't work, nor would code like "var sneaky = localStorage.getItem(…)". Just some thoughts.

**Comment by bk3c** [comment]
@jaredcwhite Yeah, I don't see any reason why conditions couldn't just resolve to a predefined list of boolean-returning functions defined on the front-end, so nothing could be injected. I agree doing an actual `eval` is a giant security hole we'd want to avoid.

I also agree with the rolling things out is phases plan, and I'd expect that even what I've put here so far could be done over a few phases. for example, I don't think we need to actually have conditionals work in the first implementation. I do think it's worthwhile to try and plan out the approach more completely, though, so we can try not to box ourselves into a corner.

**Comment by jaredcwhite** [comment]
@bk3c Cool beans. Hopefully others can weigh in on this approach soon so we can keep the momentum going.

**Comment by bk3c** [comment]
See also [this reddit thread](https://www.reddit.com/r/nextjs/comments/h108h1/how_to_proceed_with_multi_step_wizard_forms_in/) for someone else's implementation idea.

**Comment by bencpeters** [comment]
How variable do we expect this logic to be in different applications? I certainly respect the desire to have a "general" solution that can be served by the backend, but I also agree wholeheartedly that a)`eval` is a bad idea and b) expressing all these conditionals in JSON is awkward at best. I'm wondering if a hybrid approach would cover our use cases (present & near future):

The JSON defining the form defines the general order, configurable options, and presence of steps and is loaded by the conductor to start the process.

Each individual step that might have relevant conditional logic has that logic defined on the front-end (either in the component itself or in the conductor... but all the code lives with the front-end). In theory it could be even be parameterized by the JSON (e.g. a boolean flag/prop for a step where it's relevant, like `requiresLogin: true`), or it could just be baked-in "characteristic" of the step.

I think the goal would be to avoid getting so bogged down in contorting the JSON to give us a totally context-less/infinitely generalize-able framework that we end up inventing a mini DSL when a few configuration flags and agreed upon conventions/logic would serve us just as well.

**Comment by jaredcwhite** [comment]
@bencpeters Yeah, I'm of a similar mind — to me the idea would be that the JSON simply provides validation hints and certain boolean-style keys, and the step(s) know how to parse those with the conductor flow. I think anything approaching pseudo-code is going to be challenging. I've heard of other teams essentially building advanced DSLs as part of form definitions like this, only to back out after a while because it just got crazy complicated.

**Comment by james-wills-sf** [comment]
@bk3c sorry for taking so long on this, planning on taking a look at this today

File: apps/public-reference/lib/AppSubmissionContext.ts [hunk 1]
```diff
@@ -1,6 +1,50 @@
...
+  // Note: this whole function will eventually be replaced with one that reads this from the backend.
+  return {
+    listing: {}, // TODO: actual listing object
+    status: "TODO: ENUM GOES HERE",
+    languages: ["en", "es"],
+    steps: [
+      {
+        name: "Choose Language", // not user-facing, only shows in partners; user text from the step itself
+        url: "/applications/start/choose-language",
+        skipIf: { condition: "userIsLoggedIn()", skipTo: "/applications/start/what-to-expect" },
```
[the remark this answers is not in the document]
Comment (james-wills-sf) [reply apps/public-reference/lib/AppSubmissionContext.ts:14]: Additionally, if the condition function to skip the step is unlikely to change on a listing-to-listing basis it's probably simplest for it to be defined on the frontend tied to the page/frontend step definition itself. I really think the ideal is the backend just passes in a list like:
```
stepsForListing: [
  { id: 'step_id_1' },
  { id: 'step_id_2' },
  { id: 'step_id_4', props: { minAge: 65 } },
]
```

And the frontend contains the step definitions like:
```
APPLICATION_STEPS = {
  'step_id_1': {
    url: 'xyz',
    skipIf: (appData) => false,
  },
  'step_id_2': {
    url: 'xyz',
    skipIf: (appData) => false,
  },
  'step_id_3': {
    url: 'xyz',
    skipIf: (appData) => false,
  },
  'step_id_4': {
    url: 'xyz',
    requiredProps: <type object with minAge>
    skipIf: (appData) => false,
  },
}
```

I don't know enough about the practical usecases here though so some additional data from the backend may be necessary!

**Review by james-wills-sf — COMMENTED** [review]
@bk3c Thanks for starting this as just a config definition, super easy to review!

I have a few comments, in general my two recommendations are:
1. Make the steps independent of one another
2. The backend should decide what steps are shown and in what order (and maybe specifying some simple config props) but shouldn't be too entangled with the frontend code

**Comment by jaredcwhite** [comment]
Hey @james-wills-sf, appreciate your feedback. We just got a PR in to implement some of this. (https://github.com/bloom-housing/bloom/pull/676) I think some of your comments might still apply, but I want to make sure you've taken a look at that first.

**Comment by bencpeters** [comment]
We can probably close this in favor of #676? @bk3c?

**Comment by bk3c** [comment]
Yeah, I was just keeping it around for the comment thread, but let's consolidate discussion on #676.

**Pull request closed without merging**